Implement Rescaling-ACG, which is a improved version of ACG attack.

yamamura-k opened the pull request:
# Description

A new version of our ACG algorithm, Rescaling-ACG (ReACG), has been implemented. ReACG outperforms APGD and ACG, and also performs better for ImageNet models in particular. Our paper will be presented at International Conference on Pattern Recognition and Artificial Intelligence (ICPRAI), 2024.


## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Same Tests as ACG
- [x] Style check

**Test Configuration**:
- OS: Ubuntu 24.04
- Python version: 3.10.14
- ART version or commit number: 1.18.0
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] My changes have been tested using both CPU and GPU devices

beat-buesser commented:
Hi @yamamura-k Thank you very much for you pull request! I'm starting the review now, please apologise the delay.

beat-buesser commented on tests/attacks/evasion/test_rescaling_auto_conjugate_gradient.py:
```diff
@@ -0,0 +1,166 @@
+
+            x_train_mnist_adv = attack.generate(x=x_train_mnist, y=y_train_mnist)
+
+            assert np.max(np.abs(x_train_mnist_adv - x_train_mnist)) > 0.0
```
Does the attack always create the same adversarial example? Could we also test for the expected pixel values in the adversarial image?

yamamura-k replied:
A CPU-only calculation with the same initial point should produce the same adversarial example each time.
However, when using a GPU, the adversarial example may be slightly different each time because the gradient calculation on the GPU is not deterministic.

beat-buesser reviewed (changes requested):
Hi @yamamura-k The pull request looks good. Could you please take a look at the minor review comments above?

yamamura-k commented on art/attacks/evasion/rescaling_auto_conjugate_gradient.py:
```diff
@@ -0,0 +1,667 @@
+"""
+This module implements the 'Rescaling-ACG' attack.
+
+| Paper link:
```
The paper link does not appear to be publicly available yet. If it is not published after waiting for a while, we will consider uploading to arxiv.

yamamura-k commented on art/attacks/evasion/rescaling_auto_conjugate_gradient.py:
```diff
@@ -0,0 +1,667 @@
+            raise ValueError("The argument batch_size has to be of type int and larger than zero.")
+
+        # if self.loss_type not in self._predefined_losses:
+        #     raise ValueError("The argument loss_type has to be either {}.".format(self._predefined_losses))
```
I don't think these comments are necessarily necessary. They have been removed in the revised commit.

yamamura-k commented on tests/attacks/evasion/test_rescaling_auto_conjugate_gradient.py:
```diff
@@ -0,0 +1,166 @@
+
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
Thank you for your comment. This suggestion is reflected to the latest commit.

yamamura-k commented on art/attacks/evasion/rescaling_auto_conjugate_gradient.py:
```diff
@@ -0,0 +1,667 @@
+
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
Thank you for your comment. This suggestion is reflected to the latest commit.

yamamura-k commented:
@beat-buesser Thank you for your review! I have modified my codes to reflect review comments. Although I just removed redundant comments and updated some comment lines, some checks were not successful. 

Our paper is not publicly available now, so I will add a link to the paper after it becomes available.

beat-buesser commented:
@yamamura-k Thank you very much. I will fix the issue with pytest-flake8 in a separate pull request and merge this pull request afterwards.

yamamura-k commented:
@beat-buesser Thank you for your review and comments. We published our paper via arxiv, and I added the link. I think your review comments are addressed.

beat-buesser commented:
Hi @yamamura-k Thank you very much for the updates and the published paper! Please apologise the delay because of vacation and fixing a bug in the unit tests. I have updated your branch with the fixed style checks and as soon as the test run has completed I'll take a final review and merge this pull request.

beat-buesser commented:
Hi @yamamura-k Thank you very much for adding your new attack to ART! It will be part of the next release ART 1.19 around December 15 and immediately available on release branch `dev_1.19.0`.
